dimod/binary_quadratic_model.py

```diff
@@ -1606,6 +1606,88 @@ def from_json(cls, obj):
 
         return json.load(obj, object_hook=lambda d: bqm_decode_hook(d, cls=cls))
 
+    def to_bson(self):
+        """Convert a binary quadratic model to an efficient BSON format.
+
+        Returns:
+            bytes: BSON formatted bytes representing the original BQM.
+        """
+                
+
+        import bson
+
+        num_variables = len(self)
+        if num_variables > 2**16:
+            raise ValueError
+
+        variable_order = sorted(self.linear)
+        num_possible_edges = num_variables*(num_variables - 1) // 2
+        density = len(self.quadratic) / num_possible_edges
+        as_complete = density >= 0.5
+
+        lin, (i, j, _vals), off = self.to_numpy_vectors(
+            dtype=np.float32,
+            index_dtype=np.uint16,
+            sort_indices=as_complete,
+            variable_order=variable_order)
+
+        if as_complete:
+            vals = np.zeros(num_possible_edges, dtype=np.float32)
+            edge_idxs = i*(num_variables - 1) - i*(i+1)//2 + j - 1
+            vals[edge_idxs] = _vals
+
+        else:
+            vals = _vals
+
+        doc = {
+            "as_complete": as_complete,
+            "linear": bson.binary.Binary(lin.tobytes()),
+            "quadratic_vals": bson.binary.Binary(vals.tobytes()),
+            "vartype": "SPIN" if self.vartype == self.SPIN else "BINARY",
+            "offset": off,
+            "variable_order": variable_order,
+        }
+
+        if not as_complete:
+            doc["quadratic_head"] = bson.binary.Binary(i.tobytes())
+            doc["quadratic_tail"] = bson.binary.Binary(j.tobytes())
+
+        return bson.BSON.encode(doc)
+
+    @classmethod
+    def from_bson(cls, obj):
+        """Deserialize a binary quadratic model from a BSON encoding.
+
+        Args:
+            obj: bytes:
+                Byte string that represents linear and quadratic biases using 
+                the encoding used in `.to_bson`.
+        Returns:
+            :class:`.BinaryQuadraticModel`: The corresponding binary quadratic 
+            model.
+        """
+
+        import bson
+        import itertools
+
+        doc = bson.BSON.decode(obj)
+
+        lin = np.frombuffer(doc["linear"], dtype=np.float32)
+        num_variables = len(lin)
+        vals = np.frombuffer(doc["quadratic_vals"], dtype=np.float32)
+        if doc["as_complete"]:
+            i, j = list(zip(*itertools.combinations(range(num_variables), 2)))
```

Comment:
not sure you need the additional `list` call

Reply:
Correct